### ISSUE TITLE
CFP: BFD(Bidirectional Forwarding Detection) support in BGP Control plane

### ISSUE THREAD

**Issue by dylandreimerink** [cilium/cilium#22394]
## BFD(Bidirectional Forwarding Detection) support in BGP Control plane
        
BFD is a health checking mechanism made for rapid (sub second) failure detection of data forwarding planes. Its a protocol which can be employed together with protocols such as BGP, IS-IS and OSFP to shorten the time between a failure and automated actions like retracting routes to reroute traffic. The protocol is describes in [RFC5880](https://www.rfc-editor.org/rfc/rfc5880) with the addition of [RFC5881](https://www.rfc-editor.org/rfc/rfc5881) for single hop and [RFC5882](https://www.rfc-editor.org/rfc/rfc5882) for multi hop monitoring.

Adding this feature to our BGP Control Plane would allow users with peers that also support the protocol to reduce downtime in case of failures in Cilium or between Cilium and peers.

## Implementation proposal
Currently our only BGP backend is GoBGP which as of writing this has no BFD support. So we will have to implement it one way or another.

### GoBGP native
The first logical approach would be to add it to GoBGP directly via a PR. GoBGP maintainers have already [indicated](https://github.com/osrg/gobgp/issues/2563) that they are open to PRs for this feature.

Once BFD support is merged upstream and we are able to update to a GoBGP version with BFD support we will have to plumb access to it from our CRD.

### Cilium native
The second approach would be to add BFD capabilities to Cilium directly, either in the agent or the datapath or a combination of the two. BFD works for any application where forwarding takes place such as the Cilium datapath. Having a non-BGP specific implementation would allow us to also use the protocol internally to monitor inter-node connections and connections in standalone L4LB for upstream and downstream.

BFD control packets are sent over UDP in small packets (for IPv4/IPv6 single and multi hop) which is easily parse-able in both XDP and TC layers. The largest obstacle to a pure eBPF based implementation would be the need to send packets on a timer. However, if the peer is in Demand mode we only have to answer to Poll messages and optionally send echo messages back both of which are request-reply without delay.

In this case, we should manually set associated BGP sessions to AdminDown in case of a BFD failure.

### CRD plubing
No matter which route we take for BFD implementation, we will have to provide our users with a way to configure it. Since BGP Control Plane is backend agnostic and not all backends might implement BFD in future, I suggest adding a annotation for the top level CRD. Something like `io.cilium/bfd: "enabled=true,active=true,echo=false,detect-multiplied=5,receive-interval=40000,transmit-interval=20000"`

### User feedback
Any information we have regrading BFD sessions should be added to the agent output (#22009). BFD adds additional diagnostic codes to communicate why a session is down which might be useful in figuring out what is going on. We should also add BFD specific information to the peer information to show the received discriminator and timing values.

### Metrics
As soon as we have BFD session information, it should be added to the BGP metrics described in (#22010)

**Comment by AidanGG** [comment]
Thanks for writing this up. My only thoughts right now are that both the GoBGP and Cilium -native approaches have merit; GoBGP BFD for the wider Go ecosystem, and Cilium for the other non-BGP related uses that you described.

**Comment by patsevanton** [comment]
Any update?

**Comment by ArsenyBelorukov** [comment]
To have BFD in different daemon (frr for example) helps to have BGP graceful restart in cilium.
If cilium has BFD support, then BGP graceful restart will not work properly.

**Comment by vmotov** [comment]
Is this feature expected to be added in the near future?

**Comment by clearly-tyler-whitney** [comment]
I've had some degree of success running BFD as a standalone BIRD process. Obviously this comes with a litany of drawbacks which proper BFD integration with the Cilium BGP Control Plane would (ideally) avoid, but in case anyone might find it useful:

https://github.com/clearly-tyler-whitney/k8s-bfd-bird

**Comment by zeysh** [comment]
Seems that Cilium Enterprise already have BFD support: https://isovalent.com/blog/post/isovalent-enterprise-for-cilium-1-16/
But I don't know when or even if that will be available for Opensource version.

**Comment by 0dragosh** [comment]
I hope they’re not pay-gating it. Can anyone from Cilium confirm if this will be an Enterprise only feature?

**Comment by dylandreimerink** [comment]
BFD has indeed been added to Isovalent Enterprise for Cilium as an enterprise only feature.

**Comment by crabique** [comment]
So just to clarify, am I getting this right?

- There was a working BFD support with MetalLB integration
- The integration got removed in https://github.com/cilium/cilium/issues/22246 in favor of "native implementation" (explicitly mentioning BFD)
- BFD support added behind a paywall with no plans to open-source it
- ????
- Profit!

**Comment by YutaroHayakawa** [comment]
> There was a working BFD support with MetalLB integration
> The integration got removed in https://github.com/cilium/cilium/issues/22246 in favor of "native implementation" (explicitly mentioning BFD)

No, we didn't have BFD support in MetalLB integration, because it was based on the old MetalLB that don't have FRR integration.

> BFD support added behind a paywall with no plans to open-source it

Right

**Comment by lgogolin** [comment]
Any update on this ?

**Comment by beddari** [comment]
For anyone interested, in some cases with frr on a bare metal host, this could be relevant https://github.com/FRRouting/frr/pull/13060

**Comment by crytectobi** [comment]
It seems that BFD is finally getting added to upstream goBGP
https://github.com/osrg/gobgp/pull/3388
Does this change anything to get BFD for everyone in Cilium?

**Comment by kadishmal** [comment]
Native support for BFD has been added to GoBGP few weeks ago https://github.com/osrg/gobgp/pull/3402. It is available in version 4.6.0.

We can see that in Cilium 1.20.0 GoBGP was upgraded to v4.5.0. It would be nice to upgrade further and add support for BFD in Cilium.

Thanks.

**Comment by michal-rybinski** [comment]
Not sure if it was changed last minute or if @kadishmal was looking at a different code, but the released v1.20 has GoBGP v4.6.1 in it:
https://github.com/cilium/cilium/blob/450c53145dc0a872ae20dfb05598c94ef55a71e2/go.mod#L79

I am guessing then we could use this functionality, providing Cilium can surface it?
I have not seen it mentioned in any documentation so I suppose it does not atm.
Any plans in doing it or does it still stays as an Enterprise only option?
